DexPriceAggregatorUniswapV3/contracts/DexPriceAggregatorUniswapV3.sol

```diff
@@ -0,0 +1,409 @@
+        emit PoolForRouteSet(poolKey.token0, poolKey.token1, _pool);
+    }
+
+    /// @notice Fetch the Uniswap V3 pool to be queried for a tokenA:tokenB route
+    /// @param _tokenA Address of an ERC20 token contract
+    /// @param _tokenB Address of another ERC20 token contract
+    /// @return pool Address of a Uniswap V3 pool constructed with _tokenA and _tokenB
+    function getPoolForRoute(address _tokenA, address _tokenB)
+        public
+        view
+        returns (address pool)
+    {
+        return
+            _getPoolForRoute(
+                PoolAddress.getPoolKey(_tokenA, _tokenB, defaultPoolFee)
+            );
+    }
+
+    /**************************
+     * Utility view functions *
+     **************************/
+
+    /// @notice Fetch a Uniswap V3 pool's current "spot" and TWAP tick values
+    /// @param _pool Address of a Uniswap V3 pool
+    /// @param _twapPeriod Number of seconds in the past to consider for the TWAP rate
+    /// @return spotTick The pool's current "spot" tick
+    /// @return twapTick The twap tick for the last _twapPeriod seconds
+    function fetchCurrentTicks(address _pool, uint32 _twapPeriod)
+        public
+        view
+        returns (int24 spotTick, int24 twapTick)
+    {
+        spotTick = OracleLibrary.getBlockStartingTick(_pool);
+        twapTick = OracleLibrary.consult(_pool, _twapPeriod);
+    }
+
+    /// @notice Given a tick and a token amount, calculates the amount of token received in exchange
+    /// @param _tokenIn Address of an ERC20 token contract to be converted
+    /// @param _amountIn Amount of tokenIn to be converted
+    /// @param _tokenOut Address of an ERC20 token contract to convert into
+    /// @param _tick Tick value representing conversion ratio between _tokenIn and _tokenOut
+    /// @return amountOut Amount of _tokenOut received for _amountIn of _tokenIn
+    function getQuoteAtTick(
+        address _tokenIn,
+        uint128 _amountIn,
+        address _tokenOut,
+        int24 _tick
+    ) public pure returns (uint256 amountOut) {
+        return
+            OracleLibrary.getQuoteAtTick(_tick, _amountIn, _tokenIn, _tokenOut);
+    }
+
+    /// @notice Similar to getQuoteAtTick() but calculates the amount of token received in exchange
+    ///         by first adjusting into ETH
+    ///         (ie. when a route goes through an intermediary pool with ETH)
+    /// @param _tokenIn Address of an ERC20 token contract to be converted
+    /// @param _amountIn Amount of tokenIn to be converted
+    /// @param _tokenOut Address of an ERC20 token contract to convert into
+    /// @param _tick1 First tick value representing conversion ratio between _tokenIn and ETH
+    /// @param _tick2 Second tick value representing conversion ratio between ETH and _tokenOut
+    /// @return amountOut Amount of _tokenOut received for _amountIn of _tokenIn
+    function getQuoteCrossingTicksThroughWeth(
```

Comment:
- _tokenIn and _tokenOut should not be the zero address
- _tokenIn != _tokenOut
- _amountIn should be greater than zero